src/crypto/CHIPCryptoPAL.cpp

```diff
@@ -192,10 +194,111 @@ CHIP_ERROR ConvertIntegerRawToDerInternal(const ByteSpan & raw_integer, MutableB
     return CHIP_NO_ERROR;
 }
 
-} // namespace
+/**
+ * @brief Find a 4 uppercase hex digit hex value after a prefix string. Used to implement
+ *        fallback CN VID/PID encoding for PAA/PAI/DAC.
+ *
+ * @param[in] buffer - buffer in which to find the substring.
+ * @param[in] prefix - prefix to match, which must be followed by 4 uppercase hex characters
+ * @param[out] out_hex_value - on CHIP_NO_ERROR return, this will be the 16-bit hex value decoded.
+ * @return CHIP_NO_ERROR on success, CHIP_ERROR_NOT_FOUND if not detected and
+ *         CHIP_ERROR_WRONG_CERT_DN if we saw the prefix but no valid hex string.
+ */
+CHIP_ERROR Find16BitUpperCaseHexAfterPrefix(const ByteSpan & buffer, const char * prefix, uint16_t & out_hex_value)
+{
+    chip::CharSpan prefix_span = chip::CharSpan::fromCharString(prefix);
 
-namespace chip {
-namespace Crypto {
+    enum State : int
+    {
+        kFindPrefix = 0,
+        kFindHex    = 1,
+        kFoundHex   = 2,
+    };
+
+    size_t min_len_required         = prefix_span.size();
+    bool found_prefix_at_least_once = false;
+
+    for (size_t start_idx = 0; start_idx < buffer.size(); start_idx++)
+    {
+        State state = State::kFindPrefix;
+        char hex_buf[4];
+        size_t hex_len = 0;
+
+        // Make a reader starting at the new cursor space as we traverse.
+        Reader reader{ buffer };
+        reader.Skip(start_idx);
+        if (!reader.HasAtLeast(min_len_required))
+        {
+            // We possibly can't match anymore since the prefix would not be there.
+            break;
+        }
+
+        Reader prefix_reader{ reinterpret_cast<const uint8_t *>(prefix_span.data()), prefix_span.size() };
+
+        char current_char = '\0';
+        bool done         = false;
+        while (!done)
```

Comment:
The indent of this method seems quite high, consider splitting it into at least 2 since the "try_match_at_given_index" seems to be reasonably indepen=dent from the loop of trying every index possible.

Since these are on spans, would it be worth it to not use readers for most operations? It seems to me like we could `memcmp` on prefixes without having the prefix reader and state machine logic. You seem to need one size + memcmp, then one size + memcpy logic to extract the hex chars.

Reply:
The goal here was to avoid any pointer arithmetic, which is what caused the original code to be harder to understand and which caused us to miss that it didn't correctly match in some cases.

Reply:
Refactored the whole thing PTAL.